packages/plugins/hmac-upstream-signature/src/index.ts

```diff
@@ -0,0 +1,136 @@
+    query: defaultPrintFn(executionRequest.document),
+    variables: executionRequest.variables,
+  });
+export const defaultParamsSerializer = (params: GraphQLParams) =>
+  jsonStableStringify({
+    query: params.query,
+    variables: params.variables,
+  });
+
+function createCryptoKey({
+  textEncoder,
+  crypto,
+  secret,
+  usages,
+}: {
+  textEncoder: TextEncoder;
+  crypto: Crypto;
+  secret: string;
+  usages: KeyUsage[];
+}): Promise<CryptoKey> {
+  return crypto.subtle.importKey(
+    'raw',
+    textEncoder.encode(secret),
+    { name: 'HMAC', hash: 'SHA-256' },
+    false,
+    usages,
+  );
+}
+
+export function useHMACUpstreamSignature(options: HMACUpstreamSignatureOptions): MeshServePlugin {
+  const shouldSign = options.shouldSign || DEFAULT_SHOULD_SIGN_FN;
+  const extensionName = options.extensionName || DEFAULT_EXTENSION_NAME;
+  const serializeExecutionRequest =
+    options.serializeExecutionRequest || defaultExecutionRequestSerializer;
+  let key$: Promise<CryptoKey>;
+  let fetchAPI: FetchAPI;
+  let textEncoder: TextEncoder;
+
+  return {
+    onYogaInit({ yoga }) {
+      fetchAPI = yoga.fetchAPI;
+    },
+    onSubgraphExecute({ subgraphName, subgraph, executionRequest, setExecutionRequest }) {
+      if (shouldSign({ subgraphName, subgraph, executionRequest })) {
+        textEncoder ||= new fetchAPI.TextEncoder();
+        key$ ||= createCryptoKey({
+          textEncoder,
+          crypto: fetchAPI.crypto,
+          secret: options.secret,
+          usages: ['sign'],
+        });
+        return key$.then(async key => {
+          const serializedExecutionRequest = serializeExecutionRequest(executionRequest);
+          const encodedContent = textEncoder.encode(serializedExecutionRequest);
+          const signature = await fetchAPI.crypto.subtle.sign('HMAC', key, encodedContent);
+          const extensionValue = btoa(String.fromCharCode(...new Uint8Array(signature)));
+          setExecutionRequest({
+            ...executionRequest,
+            extensions: {
+              ...executionRequest.extensions,
+              [extensionName]: extensionValue,
+            },
+          });
+        });
+      }
+    },
+  };
+}
+
+export type HMACUpstreamSignatureValidationOptions = {
+  secret: string;
+  extensionName?: string;
+  serializeParams?: (params: GraphQLParams) => string;
+};
+
+export function useHMACSignatureValidation(
+  options: HMACUpstreamSignatureValidationOptions,
+): YogaPlugin {
+  const extensionName = options.extensionName || DEFAULT_EXTENSION_NAME;
+  let key$: Promise<CryptoKey>;
+  let textEncoder: TextEncoder;
+  const paramsSerializer = options.serializeParams || defaultParamsSerializer;
+
+  return {
+    onParams({ params, fetchAPI }) {
+      textEncoder ||= new fetchAPI.TextEncoder();
+      const extension = params.extensions?.[extensionName];
+      if (!extension) {
+        throw new Error(`Missing HMAC signature: extension ${extensionName} not found in request.`);
```

Comment:
We throw a raw error ? This means it will be masked, this is what we want ? It will probably make debugging a bit difficult on setup in case something went wrong somewhere

Reply:
It should hard fail in this case, if an unsigned request got to the subgraph, we want to make sure to fail it without details: it might be someone who got access to the internal network. We don't want to hint too much in this case.

Reply:
Ok makes sense :-)

Reply:
Also keep in mind, this error message is logged like any other masked errors.